reactor-core/src/main/java/reactor/core/DefaultDisposable.java

```diff
@@ -0,0 +1,417 @@
+						return true;
+					}
+				}
+			}
+			for (Disposable d : ds) {
+				d.dispose();
+			}
+			return false;
+		}
+
+		@Override
+		public boolean remove(Disposable d) {
+			Objects.requireNonNull(d, "Disposable item is null");
+			if (disposed) {
+				return false;
+			}
+			synchronized (this) {
+				if (disposed) {
+					return false;
+				}
+
+				OpenHashSet<Disposable> set = disposables;
+				if (set == null || !set.remove(d)) {
+					return false;
+				}
+			}
+			return true;
+		}
+
+		@Override
+		public void clear() {
+			if (disposed) {
+				return;
+			}
+			OpenHashSet<Disposable> set;
+			synchronized (this) {
+				if (disposed) {
+					return;
+				}
+
+				set = disposables;
+				disposables = null;
+			}
+
+			dispose(set);
+		}
+
+		@Override
+		public int size() {
+			if (disposed) {
+				return 0;
+			}
+			synchronized (this) {
+				if (disposed) {
+					return 0;
+				}
+				OpenHashSet<Disposable> set = disposables;
+				return set != null ? set.size() : 0;
+			}
+		}
+
+		/**
+		 * Dispose the contents of the OpenHashSet by suppressing non-fatal
+		 * Throwables till the end.
+		 * @param set the OpenHashSet to dispose elements of
+		 */
+		void dispose(@Nullable OpenHashSet<Disposable> set) {
+			if (set == null) {
+				return;
+			}
+			List<Throwable> errors = null;
+			Object[] array = set.keys();
+			for (Object o : array) {
+				if (o instanceof Disposable) {
+					try {
+						((Disposable) o).dispose();
+					} catch (Throwable ex) {
+						Exceptions.throwIfFatal(ex);
+						if (errors == null) {
+							errors = new ArrayList<>();
+						}
+						errors.add(ex);
+					}
+				}
+			}
+			if (errors != null) {
+				if (errors.size() == 1) {
+					throw Exceptions.propagate(errors.get(0));
+				}
+				throw Exceptions.multiple(errors);
+			}
+		}
+	}
+
+	/**
+	 * Not public API. Implementation of a {@link Disposable.SequentialDisposable}.
+	 *
+	 * @author Simon Baslé
+	 * @author David Karnok
+	 */
+	static final class AtomicSequentialDisposable implements
+	                                              Disposable.SequentialDisposable {
+
+		volatile Disposable inner;
+		static final AtomicReferenceFieldUpdater<AtomicSequentialDisposable, Disposable>
+				INNER =
+				AtomicReferenceFieldUpdater.newUpdater(AtomicSequentialDisposable.class, Disposable.class, "inner");
+
+		@Override
+		public boolean update(Disposable next) {
+			return DefaultDisposable.set(INNER, this, next);
+		}
+
+		@Override
+		public boolean replace(@Nullable Disposable next) {
+			return DefaultDisposable.replace(INNER, this, next);
+		}
+
+		@Override
+		@Nullable
+		public Disposable get() {
+			return inner;
+		}
+
+		@Override
+		public void dispose() {
+			DefaultDisposable.dispose(INNER, this);
+		}
+
+		@Override
+		public boolean isDisposed() {
+			return DefaultDisposable.isDisposed(INNER.get(this));
+		}
+	}
+
+
+	//==== STATIC ATOMIC UTILS copied from Disposables ====
+
+	static final Disposable DISPOSED = new Disposable() {
```

Comment:
@smaldini we should probably double-check that users of `SequentialDisposable` don't look for the `Disposables.DISPOSED` instance inside the sequential, instead of this one.

(both are marked as "don't leak to clients", so this is a mistake only reactor-core can make )